x/handlers.go

```diff
@@ -348,3 +357,236 @@ func (h *s3Handler) Rename(srcPath, dstPath string) error {
+	resp, err := azs.bucket.NewBlockBlobURL(azs.JoinPath(path)).Download(context.Background(), 0,
+		azblob.CountToEnd, azblob.BlobAccessConditions{}, false, azblob.ClientProvidedKeyOptions{})
+	if err != nil {
+		return nil, errors.Wrap(err, "while reading file")
+	}
+
+	buf := bytes.Buffer{}
+	if _, err := buf.ReadFrom(resp.Body(azblob.RetryReaderOptions{})); err != nil {
+		return nil, errors.Wrap(err, "while reading file")
+	}
+	return buf.Bytes(), nil
+}
+
+// Rename renames the src file to the destination file.
+func (azs *AZS) Rename(src, dst string) error {
+	ctx := context.Background()
+
+	srcHandle := azs.bucket.NewBlockBlobURL(azs.JoinPath(src))
+	resp, err := srcHandle.Download(ctx, 0,
+		azblob.CountToEnd, azblob.BlobAccessConditions{}, false, azblob.ClientProvidedKeyOptions{})
+	if err != nil {
+		return errors.Wrap(err, "while reading file")
+	}
+
+	if _, err = azblob.UploadStreamToBlockBlob(ctx, resp.Body(azblob.RetryReaderOptions{}),
+		azs.bucket.NewBlockBlobURL(azs.JoinPath(dst)),
+		azblob.UploadStreamToBlockBlobOptions{}); err != nil {
+		return errors.Wrapf(err, "while uploading")
+	}
+
+	if _, err = srcHandle.Delete(ctx, azblob.DeleteSnapshotsOptionInclude,
+		azblob.BlobAccessConditions{}); err != nil {
+		return errors.Wrapf(err, "while deleting file")
+	}
+
+	return nil
+}
+
+// Stream would stream the path via an instance of io.ReadCloser. Close must be called at the
+// end to release resources appropriately.
+func (azs *AZS) Stream(path string) (io.ReadCloser, error) {
+	resp, err := azs.bucket.NewBlockBlobURL(azs.JoinPath(path)).Download(context.Background(), 0,
+		azblob.CountToEnd, azblob.BlobAccessConditions{}, false, azblob.ClientProvidedKeyOptions{})
+	if err != nil {
+		return nil, errors.Wrap(err, "while reading file")
+	}
+
+	return resp.Body(azblob.RetryReaderOptions{}), nil
+}
+
+type azWriter struct {
+	pReader *io.PipeReader
+	pWriter *io.PipeWriter
+	errCh   chan error
+}
+
+// Write writes to the pipe writer.
+func (w *azWriter) Write(p []byte) (n int, err error) {
+	return w.pWriter.Write(p)
+}
+
+// Close calls close on the pipe writer and returns any erorrs encoutered in writing using the
+// pipe reader.
+func (w *azWriter) Close() error {
+	if w == nil {
+		return nil
+	}
+
+	if err := w.pWriter.Close(); err != nil {
+		glog.Errorf("Unexpected error when closing pipe: %v", err)
+	}
+	w.pWriter = nil
+
+	return <-w.errCh
+}
+
+// Helper function to process writes to azure. The function is run is a separate go-routine because
+// the azure API requires a reader instead of a writer as an input. Any errors are returned to
+// errCh of the azWriter, which are returned when close on the azWriter is called.
+func (w *azWriter) upload(bucket *azblob.ContainerURL, absPath string) {
+	f := func() error {
+		ctx := context.Background()
+		_, err := azblob.UploadStreamToBlockBlob(ctx, w.pReader, bucket.NewBlockBlobURL(absPath),
+			azblob.UploadStreamToBlockBlobOptions{})
+
+		if err != nil {
```

Comment:
No need for this if statement. 

You can directly ` return errors.Wrapf(err, ...)`.